sdks/java/io/thrift/src/main/java/org/apache/beam/sdk/io/thrift/ThriftIO.java

```diff
@@ -0,0 +1,289 @@
+ * FileIO.ReadableFile}, use the {@link ReadFiles} transform.
+ *
+ * <p>For example:
+ *
+ * <pre>{@code
+ * PCollection<FileIO.ReadableFile> files = pipeline
+ *   .apply(FileIO.match().filepattern(options.getInputFilepattern())
+ *   .apply(FileIO.readMatches());
+ *
+ * PCollection<ExampleType> examples = files.apply(ThriftIO.readFiles(ExampleType.class).withProtocol(thriftProto);
+ * }</pre>
+ *
+ * <h3>Writing Thrift Files</h3>
+ *
+ * <p>{@link ThriftIO.Sink} allows for a {@link PCollection} of {@link TBase} to be written to
+ * Thrift files. It can be used with the general-purpose {@link FileIO} transforms with
+ * FileIO.write/writeDynamic specifically.
+ *
+ * <p>For example:
+ *
+ * <pre>{@code
+ * pipeline
+ *   .apply(...) // PCollection<ExampleType>
+ *   .apply(FileIO
+ *     .<ExampleType>write()
+ *     .via(ThriftIO.sink(thriftProto))
+ *     .to("destination/path");
+ * }</pre>
+ *
+ * <p>This IO API is considered experimental and may break or receive backwards-incompatible changes
+ * in future versions of the Apache Beam SDK.
+ */
+@Experimental(Experimental.Kind.SOURCE_SINK)
+public class ThriftIO {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ThriftIO.class);
+
+  /** Disable construction of utility class. */
+  private ThriftIO() {}
+
+  /**
+   * Reads each file in a {@link PCollection} of {@link org.apache.beam.sdk.io.FileIO.ReadableFile},
+   * which allows more flexible usage.
+   */
+  public static <T> ReadFiles<T> readFiles(Class<T> recordClass) {
+    return new AutoValue_ThriftIO_ReadFiles.Builder<T>().setRecordClass(recordClass).build();
+  }
+
+  //////////////////////////////////////////////////////////////////////////////
+
+  /** Creates a {@link Sink} for use with {@link FileIO#write} and {@link FileIO#writeDynamic}. */
+  public static <T extends TBase<?, ?>> Sink<T> sink(TProtocolFactory factory) {
+    return new AutoValue_ThriftIO_Sink.Builder<T>().setProtocolFactory(factory).build();
+  }
+
+  //////////////////////////////////////////////////////////////////////////////
+
+  /** Implementation of {@link #readFiles}. */
+  @AutoValue
+  public abstract static class ReadFiles<T>
+      extends PTransform<PCollection<FileIO.ReadableFile>, PCollection<T>> {
+
+    abstract Builder<T> toBuilder();
+
+    @Nullable
+    abstract Class<T> getRecordClass();
+
+    @Nullable
+    abstract TProtocolFactory getTProtocolFactory();
+
+    public ReadFiles<T> withProtocol(TProtocolFactory protocol) {
+      return toBuilder().setTProtocolFactory(protocol).build();
+    }
+
+    @Override
+    public PCollection<T> expand(PCollection<FileIO.ReadableFile> input) {
+      checkNotNull(getRecordClass(), "Record class cannot be null");
+      checkNotNull(getTProtocolFactory(), "Thrift protocol cannot be null");
+
+      final Coder<T> coder = ThriftCoder.of();
+      return input
+          .apply(ParDo.of(new ReadFn<>(getRecordClass(), getTProtocolFactory())))
+          .setCoder(coder);
```

Comment:
Actually you can get rid of the ThriftCoder class by relying on standard Java serialization doing this here:
`.setCoder(SerializableCoder.of(new TypeDescriptor() {}));`
But hopefully we could use 'native' Thrift Coder.